core/tests/test_utils.py

```diff
@@ -1245,7 +1245,199 @@ def assert_matches_regexps(self, items, regexps, full_match=False):
+
+    def get_pending_mapreduce_tasks(self, queue_name=None):
+        """Returns the jobs in the given mapreduce taskqueue. If queue_name is
+        None, defaults to returning the jobs in all available queues.
+        """
+        queue_names = None if queue_name is None else [queue_name]
+        return self._taskqueue_stub.get_filtered_tasks(queue_names=queue_names)
+
+    def _execute_mapreduce_tasks(self, tasks):
+        """Execute mapreduce queued tasks.
+
+        Args:
+            tasks: list(google.appengine.api.taskqueue.taskqueue.Task). The
+                queued tasks.
+        """
+        for task in tasks:
+            if task.url == '/_ah/queue/deferred':
+                deferred.run(task.payload)
+            else:
+                # All other tasks are expected to be for mapreduce or taskqueue.
+                params = task.payload or ''
+                headers = {
+                    'Content-Length': python_utils.convert_to_bytes(len(params))
+                }
+                headers.update(
+                    (key, python_utils.convert_to_bytes(val))
+                    for key, val in task.headers.items())
+
+                app = (
+                    self.taskqueue_testapp if task.url.startswith('/task') else
+                    self.testapp)
+                response = app.post(
+                    task.url, params=params, headers=headers,
+                    expect_errors=True)
+                if response.status_code != 200:
+                    raise RuntimeError('MapReduce task failed: %r' % task)
+
+    def process_and_flush_pending_mapreduce_tasks(self, queue_name=None):
+        """Runs and flushes pending mapreduce tasks. If queue_name is None, does
+        so for all queues; otherwise, this only runs and flushes tasks for the
+        specified queue.
+
+        For more information on taskqueue_stub, see:
+        https://code.google.com/p/googleappengine/source/browse/trunk/python/google/appengine/api/taskqueue/taskqueue_stub.py
+        """
+        queue_names = (
+            self._get_all_queue_names() if queue_name is None else [queue_name])
+
+        get_enqueued_tasks = lambda: list(
+            self._taskqueue_stub.get_filtered_tasks(queue_names=queue_names))
+
+        # Loop until get_enqueued_tasks() returns an empty list.
+        for tasks in iter(get_enqueued_tasks, []):
+            for queue in queue_names:
+                self._taskqueue_stub.FlushQueue(queue)
+            self._execute_mapreduce_tasks(tasks)
+
+    def run_but_do_not_flush_pending_mapreduce_tasks(self):
+        """"Runs but not flushes mapreduce pending tasks."""
+        queue_names = self._get_all_queue_names()
+
+        tasks = self._taskqueue_stub.get_filtered_tasks(queue_names=queue_names)
+        for queue in queue_names:
+            self._taskqueue_stub.FlushQueue(queue)
+        self._execute_mapreduce_tasks(tasks)
+
+
+class GenericTestBase(AppEngineTestBase):
```

Comment:
How does one choose to use this or the more minimal class? It seems like GenericTestBase can always be used since it's inheriting the behavior of AppEngineTestBase.

Maybe try to make it really clear in the docstring when to choose one versus the other to help guide people better. That being said, given that a lot of people just copy/paste tests I suspect we'll see a lot of unnecessary usage of the heavierweight class.

A separate aside: this is a situation where composition would be much more preferable over inheritance since then individual tests can decide whether to decorate their suite with the additional behavior that's desired. However, it seems that Python unittest doesn't well support composition-based extensions, so I don't actually have any specific suggestion here. This might be something to consider improving in the future, so I wanted to mention it as a potential gap.

Reply:
OK, I've tried to explain this in the improved docstring, what do you think?

/cc @seanlip

Reply:
Also, regarding your aside, I think mixins would be an elegant solution and agree that composition would be much more preferable. However, that will likely need a lot of work so I don't want to invest too much effort yet.

Reply:
Just noting that the docstring above looks better now, thanks @brianrodri! 

This one is a bit sparse, though, with regards to usage details. Can you maybe add a note to it saying something like "All tests outside core/platform, or that are not for platform-specific functionality, should inherit from this base class"?

Reply:
Hmm, `auth_jobs_one_off_test` needs to inherit from `AppEngineTestBase`, so that rule is probably too strict for now. I don't think we should "enforce" anything with comments until we come up with an "attack plan" for breaking up GenericTestBase (which we can discuss in #12135).

Reply:
Maybe just add language like "typically" here, then? And for any files that break the rule, add an explanation why.

That will help give guidance without being 100% prescriptive.

Reply:
OK, I've expanded on it. WDYT?

Reply:
It's good (from my end). Thanks!